api/client/src/main/java/org/projectnessie/client/http/HttpClientBuilderImpl.java

```diff
@@ -240,32 +252,56 @@ public HttpClient build() {
             .addAllResponseFilters(responseFilters)
             .isHttp11Only(!http2Upgrade)
             .followRedirects(followRedirects)
-            .forceUrlConnectionClient(forceUrlConnectionClient)
             .authentication(authentication)
             .build();
 
-    return ImplSwitch.FACTORY.apply(config);
+    String clientName = httpClientName;
+    if ("HTTP".equalsIgnoreCase(clientName)) {
+      // fall back to default
+      clientName = null;
+    }
+    if (clientName == null && forceUrlConnectionClient) {
+      // forced use of URLConnectionClient
+      clientName = "URLConnection";
+    }
+    if (clientName != null
+        && forceUrlConnectionClient
+        && !"URLConnection".equalsIgnoreCase(clientName)) {
+      throw new IllegalArgumentException(
+          "Both forceUrlConnectionClient and httpClientName are specified with incompatible values, migrate to httpClientName");
+    }
+
+    HttpClientFactory httpClientFactory =
+        clientName != null
+            ? ImplSwitch.IMPLEMENTATIONS.get(clientName.toLowerCase(ROOT))
+            : ImplSwitch.PREFERRED;
+    if (httpClientFactory == null) {
+      throw new IllegalArgumentException(
+          "No HTTP client factory for name '" + clientName + "' found");
+    }
+    return httpClientFactory.buildClient(config);
+  }
+
+  static Set<String> clientNames() {
+    return ImplSwitch.IMPLEMENTATIONS.keySet();
   }
 
   static class ImplSwitch {
-    static final Function<HttpRuntimeConfig, HttpClient> FACTORY;
+    static final Map<String, HttpClientFactory> IMPLEMENTATIONS;
+    static final HttpClientFactory PREFERRED;
 
     static {
-      Function<HttpRuntimeConfig, HttpClient> factory;
-      try {
-        Class.forName("java.net.http.HttpClient");
-        factory =
-            config ->
-                // Need the system property for tests, "normal" users can use standard
-                // configuration options.
-                Boolean.getBoolean("nessie.client.force-url-connection-client")
-                        || config.forceUrlConnectionClient()
-                    ? new UrlConnectionClient(config)
-                    : new JavaHttpClient(config);
-      } catch (ClassNotFoundException e) {
-        factory = UrlConnectionClient::new;
+      IMPLEMENTATIONS = new HashMap<>();
+      HttpClientFactory preferred = null;
+      for (HttpClientFactory s : ServiceLoader.load(HttpClientFactory.class)) {
+        if (s.available()) {
+          if (preferred == null || preferred.priority() > s.priority()) {
+            preferred = s;
+          }
+          IMPLEMENTATIONS.put(s.name().toLowerCase(ROOT), s);
+        }
       }
-      FACTORY = factory;
+      PREFERRED = preferred;
```

Comment:
Should we log a warning or throw if `preferred` is null?

Reply:
OK let's assume it will never happen since at least `URLConnection` will be available. Nevermind.